[docker] support for different compile profiles and features

rustielin opened the pull request:
### Description

Implement the ability to build/tag/push/cache our rust images while being able to specify a cargo profile and list of optional features. This allows us to more easily and continuously test the following scenarios:
* separate `aptos/validator-testing` image that includes a bunch of testing and debugging goodies developers might want to use manually, e.g. in testnets or Forge. The full list of packages is at the end of the dockerfile
* use `failpoints` feature for consensus tests
* benchmarking and testing with `performance` builds

To ensure the docker layer caches don't clobber each other, since we're essentially building multiple times for the same GIT_SHA, use a series of docker tag prefixes. `release` images with the default (no additional features) build config will continue to be tagged `GIT_SHA` as expected. Images of other profiles will be tagged `<PROFILE>_<GIT_SHA>`, and if there are additional features, then `<PROFILE>_<FEATURES>_<GIT_SHA>`. Docker tags can be 128 characters, so there's potential of running over that if we specify a long list of features, but for now, it should be fine to support the `failpoints` use case.

These builds might be more expensive, so we can build them continuously on `main` (e.g. every hour) rather than on PR. For now though, they will trigger on the same events as the normal rust build, and be non-land-blocking.

Also, include in the Forge python wrapper the ability to specify `FORGE_ENABLE_FAILPOINTS`, which will do the same tests as before, but in the image selection logic, the image tags will implicitly be prefixed with `failpoints_`. The logic for replacing the image name for the new test images is a bit more deeply embedded in the Forge rust code, so I'll tackle that in a separate PR. 

### Test Plan

Because of `pull_request_target`.... test by pushing to this branch, which is a temporary build trigger. Then check the following:
* Builds succeed: https://github.com/aptos-labs/aptos-core/actions/runs/2959605878/attempts/1
* Images are tagged as expected, with no overlap in cache keys / image tags. For example, we have created these images as a result of canary:
  * `aptos/validator:performance_16882a13ab86cbcb407a3256c5879ba5a2ad4009`
  * `aptos/validator:failpoints_16882a13ab86cbcb407a3256c5879ba5a2ad4009`
  * `aptos/validator-testing:failpoints_16882a13ab86cbcb407a3256c5879ba5a2ad4009`
* Re-triggering the build on the same branch should speed up all builds, as each has its own remote docker layer cache: https://github.com/aptos-labs/aptos-core/actions/runs/2959605878/attempts/3

For the Forge wrapper, wrote a new unit test

<!-- Please provide us with clear details for verifying that your changes work. -->

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/3538)
<!-- Reviewable:end -->

zekun000 commented:
the rust-images in last run still takes 30mins?

sitalkedia commented on .github/workflows/build-images.yaml:
```diff
@@ -123,6 +125,70 @@ jobs:
           LAST_GREEN_COMMIT: ${{ needs.determine-last-green-commit.outputs.lastGreenCommit }}
 
+  # TODO(rustielin): build these images continuously rather than on PR
+  rust-images-testing:
```
Will this be a land blocker?

rustielin replied:
> Will this be a land blocker?

These are not land-blocking. It will run on the same triggers for now (we can move to continuous later) though.

> How would we specify the image tag in the tests? Do we have that support already?

You would have to know the image tag beforehand. If you have the GIT_SHA, you can do like `<profile>_<features>_GIT_SHA`

igor-aptos replied:
how would I use this diff to create a continuous test that uses failpoints?

rustielin replied:
You would need to specify an `IMAGE_TAG` yourself of the format I wrote above (you can either look at the github actions runs or the list of builds in our image registries), or we can bake into the forge invocation script a flag to select these images automatically. 

The trick is that the image tags all follow a standard format, so if we know what git revision to test, we can programmatically select the image build that has failpoints

igor-aptos replied:
I am not asking for individual runs,  I'm asking for continuous test.  So every run would be on a different IMAGE_TAG?

So can we bake this in forge invocation so that I can configure a test always to use inage with failpoints?

rustielin replied:
Oh I see what you mean -- yea if the image is different for each run, you'd need a way to toggle failpoints. Let me push up something for that

rustielin replied:
Ok just updated the wrapper script to handle `FORGE_ENABLE_FAILPOINTS`

sitalkedia reviewed (commented):
@rustielin - How would we specify the image tag in the tests? Do we have that support already?

rustielin commented:
Here's a few successive funs. Github UI is a bit wonky and you need to select different "Attempt #"s for each run:
* attempt 1: https://github.com/aptos-labs/aptos-core/actions/runs/2959605878/attempts/1
* attempt 2: https://github.com/aptos-labs/aptos-core/actions/runs/2959605878/attempts/2
* attempt 3: https://github.com/aptos-labs/aptos-core/actions/runs/2959605878/attempts/3

rustielin commented:
Testing again after the cache optimization PR landed: https://github.com/aptos-labs/aptos-core/pull/3561

* First run takes a while: https://github.com/aptos-labs/aptos-core/actions/runs/2967777857/attempts/1 (the main `rust-images` was already cached from a previous run)
* Second run uses cache: https://github.com/aptos-labs/aptos-core/actions/runs/2967777857/attempts/2

perryjrandall commented on testsuite/forge.py:
```diff
@@ -942,33 +942,40 @@ def find_recent_images(
     num_images: int,
     # Set a generoush threshold in case of failures
     commit_threshold: int = 100,
+    enable_failpoints: bool = False,
```
lets pass image_name here that decouples this logic from the intent of failpoints

rustielin replied:
My thinking here was that it's easier to test if we have `enable_failpoints` and all these special cases in the function itself, since the logic might get more complex.

perryjrandall replied:
I feel pretty strongly this is a violation of this interface, if you need to add more logic lets introduce it in another function and leave this function just searching for images, when we couple too much specific logic here it becomes unusable elsewhere. I probably should have passed in image_tag from the start :P

perryjrandall commented on testsuite/forge.py:
```diff
@@ -1092,17 +1101,21 @@ def test(
-    default_latest_image, second_latest_image = list(find_recent_images(shell, git, 2))
     if forge_test_suite == "compat":
+        # Compat uses 2 image tags
+        default_latest_image, second_latest_image = list(find_recent_images(shell, git, 2, enable_failpoints=forge_enable_failpoints))
```
following the comment above, we could just add image tag here 

```
image_tag = "failpoints_aptos/validator" if forge_enable_failpoints else  "aptos/validator"
```

perryjrandall commented on testsuite/forge.py:
```diff
@@ -1111,6 +1124,11 @@ def test(
     assert forge_image_tag is not None, "Forge image tag is required"
     assert upgrade_image_tag is not None, "Upgrade image tag is required"
 
+    print("Using the following image tags:")
```
should we add some of this in the report message?

rustielin replied:
The comment output will print the image tag the test used, and in the case of compat, the test itself injects test report text regarding the order of upgrades as they roll out to the network

perryjrandall commented on testsuite/forge_test.py:
```diff
@@ -232,6 +232,16 @@ def testFindRecentImage(self) -> None:
+        ]))
+        git = Git(shell)
+        image_tags = find_recent_images(shell, git, 1, enable_failpoints=True)
+        self.assertEqual(list(image_tags), ["failpoints_tomato"])
```
excellent test naming

perryjrandall commented on .github/workflows/build-images.yaml:
```diff
@@ -125,6 +127,74 @@ jobs:
           LAST_GREEN_COMMIT: ${{ needs.determine-last-green-commit.outputs.lastGreenCommit }}
 
+  # TODO(rustielin): build these images continuously rather than on PR
+  rust-images-testing:
```
Can we abstract the image build into a sub workflow and call it twice here?

Then we can actually use it more simply elsewhere too

perryjrandall commented on testsuite/forge.py:
```diff
@@ -994,6 +1001,7 @@ def sanitize_forge_namespace(forge_namespace: str) -> str:
 @envoption("FORGE_NAMESPACE_KEEP")
 @envoption("FORGE_NAMESPACE_REUSE")
 @envoption("FORGE_ENABLE_HAPROXY")
+@envoption("FORGE_ENABLE_FAILPOINTS")
```
How does this option actually get triggered?

It seems like the failpoints might depend on the actual test suite being called?

rustielin replied:
For now, the test suite that uses failpoints must be triggered with `FORGE_ENABLE_FAILPOINTS`. I imagine there's a better way to wrap it once the suite is actually written, the same way we have some special logic for the `compat` test suite as well

perryjrandall reviewed (changes requested):
Overall looking awesome, just a couple things i see

1) I dont understand how forge is actually using the FORGE_FAILPOINT value? Does it depend on the test being executed?
2) The forge invocation doesnt actually wait on the image build to finish, its likely that forge can run before one of the images finish
2) Can we refactor the actual image build to just build an image to avoid duplicating that logic? That would be pretty slick, because "build-images" actually does a lot more than build images perhaps we should rename that to what it actually does like "rust-e2e-test" or something like that

rustielin commented:
Thanks for the thorough review @perryjrandall ! Addressing your comments:

> I dont understand how forge is actually using the FORGE_FAILPOINT value? Does it depend on the test being executed?

Right now it's dumb and requires whoever's invoking the test to set `FORGE_ENABLE_FAILPOINTS=true` along with the test suite. The test hasn't been written yet, so it's just easier to test this way rn. Once the test is written, we can do something like "if test_suite = consensus_failpoints_test, then use these special images", but I'm thinking we do that later

> The forge invocation doesnt actually wait on the image build to finish, its likely that forge can run before one of the images finish

Yea that's by design for now. We won't be running these failpoints tests as land-blocking, so they're just parallel builds that we just happen to kick off on PRs. If we want to switch to just doing them continuously later (either due to long or expensive build times), then we can do that. Then once the tests are written that rely on these images, we can either rely on the same git rev searching logic that we have already (just for a slightly different image tag), or just simply build the images before the continuous test in the same workflow

> Can we refactor the actual image build to just build an image to avoid duplicating that logic? That would be pretty slick, because "build-images" actually does a lot more than build images perhaps we should rename that to what it actually does like "rust-e2e-test" or something like that

yea let me look into that. seems reasonable

perryjrandall commented on .github/workflows/docker-build-test.yaml:
```diff
@@ -94,39 +96,53 @@ jobs:
-        env:
-          LAST_GREEN_COMMIT: ${{ needs.determine-last-green-commit.outputs.lastGreenCommit }}
+  rust-images:
+    needs: [permission-check, determine-last-green-commit, determine-docker-build-metadata]
```
This is super clean :D and very reusable!

perryjrandall commented on .github/workflows/docker-rust-build.yaml:
```diff
@@ -0,0 +1,66 @@
+name: "Build+Push Rust Docker Images"
```
Actually now i can even use this in my exp build script too :D

perryjrandall reviewed (approved):
Ah thanks for detailed explanation accepting with two requests

1) Lets block on all the image builds, either build them all the time and depend on them or dont build them at all

2) I feel pretty strongly about that find_image_function interface, lets try to keep those functions as loosely coupled as possible, it makes them easier to read and test and allows us to build more complex things on top of them

igor-aptos commented:
My understanding is that,  whichever continuous test needs failpoints will be able to specify that in its yaml file (like .github/workflows/continuous-e2e-consensus-fault-tolerance.yaml)

If that's how this works,  that seems best (even better than hardcoding in forge). I don't understand a lot of the code here, but it looks good overall,  and will add a test today,  so that we can try it out
